### PR TITLE
really return on context done

### PR DESCRIPTION
**Pull request by akademic**
Hi, Axel

I noticed that the comment for the function doesn't match its operation. 
So, I fixed the function to work as described in the comment.

### CODE REVIEW

**Comment by Merovius** [comment]
Alternatively, if that doesn't work for some reason, another possibility that might work is spawning a goroutine that waits for the context to be done and then either calls `l.SetDeadline(time.Now())` or just closes it.

I think ideally there should be a test case for this, just to reproduce it and make sure the fix actually fixes it.

**Comment by akademic** [comment]
> seems like a hack

I agree. My first implementation was dirty - did it in the night to get fast result in my task.
I tested approach with [(*net.ListenConfig).Listen](https://pkg.go.dev/net#ListenConfig.Listen)
It does not work as may expect - it does not finish l.Accept waiting on context.Done.

So I made new version inspired by this answer https://stackoverflow.com/a/66755998

Check if it suitable for you, or let me know how can I improve this PR

**Comment by Merovius** [comment]
Yes, it seems this is golang/go#10527. Hm.

I don't think this is quite correct. If you call this with a context that never gets cancelled and then the `Accept` call fails for a *different* reason (e.g. the network goes down), this will leak a goroutine. Also, in general, a `select` with a `default` clause is a red flag. In this case, you don't need the `quit` channel, as you can use `errors.Is(err, net.ErrClosed)` to distinguish the case where the listener was closed.

I think the steps needed are

1. Use `ctx, cancel := context.WithCancel(ctx); defer cancel()` to get a child context that definitely gets cancelled eventually. Do that after the `defer wg.Done()`, so the cancellation happens *before* we wait on the `WaitGroup`. i.e. move the `WaitGroup` stuff to the very top of the function.
2. Put the goroutine that does the closing into the same `WaitGroup`. We want to guarantee that the listener is closed before returning. This is also why it's important to `defer wg.Done()` before the `defer cancel()`.

**Comment by akademic** [comment]
Here are fixes you requested

File: transmission.go [hunk 1]
```diff
@@ -48,15 +48,30 @@ type Device interface {
...
 		return err
 	}
-	var wg sync.WaitGroup
-	defer wg.Wait()
+
+	wg.Add(1)
+	go func() {
+		<-ctxCancel.Done()
+		l.Close()
+		wg.Done()
```
Comment (Merovius) [inline transmission.go:66]: Please `defer wg.Done()` at the top of the function. It's unlikely to make a semantic difference, but it's better to stay with the consistent pattern.

File: transmission_test.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	defer cancel()
+
+	dir, _ := os.MkdirTemp("", "")
+
+	sockFile := filepath.Join(dir, "nbd.sock")
+
+	// Start the server
+	exited := make(chan any)
+	go func() {
+		lErr := ListenAndServe(ctx, "unix", sockFile, Export{})
```
Comment (Merovius) [inline transmission_test.go:22]: Just call this `err`.

File: transmission_test.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	case <-exited:
+		// The server context should be cancelled, and the server should shut down, within a very short time.
+	}
+}
+
+func TestListenAndServeContextNoCancel(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dir, _ := os.MkdirTemp("", "")
```
Comment (Merovius) [inline transmission_test.go:47]: `dir := t.TempDir()`

File: transmission_test.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+	sockFile := filepath.Join(dir, "nbd.sock")
+
+	// Start the server
+	exited := make(chan any)
+	go func() {
+		lErr := ListenAndServe(ctx, "unix", sockFile, Export{})
```
Comment (Merovius) [inline transmission_test.go:54]: Just `err`

File: transmission_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		if lErr != nil {
+			t.Errorf("ListenAndServe returned an error: %v", lErr)
+		}
+		close(exited)
```
Comment (Merovius) [inline transmission_test.go:26]: Make this `defer close(exited)` at the top of the function (guard against `ListenAndServe` panicing in a future change).

File: transmission_test.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
```
Comment (Merovius) [inline transmission_test.go:58]: `defer close(exited)` at the top here as well.

File: transmission_test.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	"path/filepath"
+	"testing"
+	"time"
+)
+
+func TestListenAndServeContextCancel(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dir, _ := os.MkdirTemp("", "")
```
Comment (Merovius) [inline transmission_test.go:15]: `dir := t.TempDir()`

**Review by Merovius — COMMENTED** [review]
Apologies for the delays. 

Overall this looks good, but there are a couple of nits. I realize that it is a bit strange to become nitpicky, given that the rest of the code is probably not entirely tidy (I assume, given how old it is). But better late, than never. I really appreciate your contribution.

File: transmission.go (hunk 1 above)
Comment (akademic) [reply transmission.go]: ok

File: transmission_test.go (hunk 2 above)
Comment (akademic) [reply transmission_test.go]: ok

File: transmission_test.go (hunk 3 above)
Comment (akademic) [reply transmission_test.go]: Didn't know about this. Thanx a lot, it will make my tests better.

File: transmission_test.go (hunk 4 above)
Comment (akademic) [reply transmission_test.go]: ok

File: transmission_test.go (hunk 5 above)
Comment (akademic) [reply transmission_test.go]: Good point

File: transmission_test.go (hunk 6 above)
Comment (akademic) [reply transmission_test.go]: ok

File: transmission_test.go (hunk 7 above)
Comment (akademic) [reply transmission_test.go]: ok

**Comment by akademic** [comment]
Here commits with changes after review.
As for code quality, my philosophy is that if project works better after patch, then everything goes ok

**Comment by Merovius** [comment]
Thanks again.

**Pull request merged**